subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/CompilationSourceDirs.java

```diff
@@ -38,32 +42,35 @@
 @NonNullApi
 public class CompilationSourceDirs {
     private static final Logger LOG = LoggerFactory.getLogger(CompilationSourceDirs.class);
+    private final List<File> sourceRoots;
 
-    private final FileTreeInternal sources;
-    private SourceRoots sourceRoots;
-
-    public CompilationSourceDirs(FileTreeInternal sources) {
-        this.sources = sources;
-    }
-
-    public List<File> getSourceRoots() {
-        return resolveRoots().getSourceRoots();
+    public CompilationSourceDirs(List<File> sourceRoots) {
+        this.sourceRoots = sourceRoots;
     }
 
-    public boolean canInferSourceRoots() {
-        return resolveRoots().isCanInferSourceRoots();
+    public static List<File> inferSourceRoots(FileTreeInternal sources) {
+        SourceRoots visitor = new SourceRoots();
+        sources.visitLeafCollections(visitor);
+        return visitor.canInferSourceRoots ? visitor.sourceRoots : Collections.emptyList();
     }
 
-    private SourceRoots resolveRoots() {
-        if (sourceRoots == null) {
-            SourceRoots visitor = new SourceRoots();
-            sources.visitLeafCollections(visitor);
-            sourceRoots = visitor;
+    /**
+     * Calculate the relative path to the source root.
+     */
+    public Optional<File> relativize(File sourceFile) {
+        if (sourceFile.isAbsolute()) {
+            return sourceRoots.stream()
+                .filter(sourceDir -> sourceFile.getAbsolutePath().startsWith(sourceDir.getAbsolutePath()))
+                .map(sourceDir -> RelativePathUtil.relativePath(sourceDir, sourceFile))
+                .filter(relativePath -> !relativePath.startsWith(".."))
+                .map(File::new)
+                .findFirst();
+        } else {
+            return Optional.of(sourceFile);
         }
-        return sourceRoots;
     }
 
-    private static class SourceRoots implements FileCollectionLeafVisitor {
+    private static class SourceRoots implements FileCollectionLeafVisitor, Serializable {
```

Comment:
Why does `SourceRoots` need to be serializable?

Reply:
It was an accident. Removed.